File: android/src/main/java/com/github/chenqingze/wechatsdk/WechatSDKPlugin.java
package com.github.chenqingze.wechatsdk;

import static com.github.chenqingze.wechatsdk.Constants.ERROR_INVALID_PARAMETERS;
import static com.github.chenqingze.wechatsdk.Constants.ERROR_SEND_REQUEST_FAILED;
import static com.github.chenqingze.wechatsdk.Constants.ERROR_WECHAT_NOT_INSTALLED;

import android.content.Context;
import android.graphics.Bitmap;
import android.graphics.BitmapFactory;
import android.os.AsyncTask;
import android.util.Log;

import com.getcapacitor.Bridge;
import com.getcapacitor.JSObject;
import com.getcapacitor.Plugin;
import com.getcapacitor.PluginCall;
import com.getcapacitor.PluginMethod;
import com.getcapacitor.annotation.CapacitorPlugin;
import com.tencent.mm.opensdk.modelbiz.WXLaunchMiniProgram;
import com.tencent.mm.opensdk.modelmsg.SendAuth;
import com.tencent.mm.opensdk.modelmsg.SendMessageToWX;
import com.tencent.mm.opensdk.modelmsg.WXImageObject;
import com.tencent.mm.opensdk.modelmsg.WXMediaMessage;
import com.tencent.mm.opensdk.modelmsg.WXMiniProgramObject;
import com.tencent.mm.opensdk.modelmsg.WXTextObject;
import com.tencent.mm.opensdk.modelmsg.WXWebpageObject;
import com.tencent.mm.opensdk.modelpay.PayReq;
import com.tencent.mm.opensdk.openapi.IWXAPI;
import com.tencent.mm.opensdk.openapi.WXAPIFactory;

@CapacitorPlugin(name = "WechatSDK")
public class WechatSDKPlugin extends Plugin {

    private WechatSDK implementation = new WechatSDK();

    public static Bridge bridge;
    public static IWXAPI wxApi;
    public static String callbackId;

    /**
     * 初始化操作
     */
    @Override
    public void load() {
        super.load();
        bridge = this.getBridge();
        registerWeChat();
    }


    /**
     * 测试
     *
     * @param call
     */
    @PluginMethod
    public void echo(PluginCall call) {
        String value = call.getString("value");
        JSObject ret = new JSObject();
        ret.put("value", implementation.echo(value));
        call.resolve(ret);
    }


    /**
     * 微信是否安装:todo ios改成未安装时改为reject
     *
     * @param call
     */
    @PluginMethod
    public void isInstalled(PluginCall call) {

        if (wxApi.isWXAppInstalled()) {
            call.resolve();
        } else {
            call.reject(ERROR_WECHAT_NOT_INSTALLED);
        }

    }


    /**
     * 支付
     *
     * @param call
     */
    @PluginMethod
    public void pay(PluginCall call) {

        bridge = this.getBridge();
        //        requestPermissions();
        PayReq req = new PayReq(); // appId
        req.appId = getWxAppId();
        req.partnerId = this.getMchId(); // 商户号
        req.prepayId = call.getString("prepayId"); // 预支付交易会话标识
        req.nonceStr = call.getString("nonceStr"); // 随机字符串
        req.timeStamp = call.getString("timeStamp"); // 时间戳
        req.packageValue = call.getString("packageValue"); // 签名
        req.sign = call.getString("sign");// 签名

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }

        callbackId = call.getCallbackId();
        if (callbackId != null) {
            Log.d(Constants.TAG, "print===================>saveCall");
            bridge.saveCall(call);
        }

        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }

    /**
     * 分享文本
     *
     * @param call
     */
    @PluginMethod
    public void shareText(PluginCall call) {
        Integer scene = call.getInt("scene");
        String text = call.getString("text");
        //初始化一个 WXTextObject 对象，填写分享的文本内容
        WXTextObject textObj = new WXTextObject();
        textObj.text = text;

        //用 WXTextObject 对象初始化一个 WXMediaMessage 对象
        WXMediaMessage msg = new WXMediaMessage();
        msg.mediaObject = textObj;
        msg.description = text;

        SendMessageToWX.Req req = new SendMessageToWX.Req();
        req.transaction = buildTransaction("text");
        req.message = msg;

        req.scene = scene;

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        callbackId = call.getCallbackId();
        if (callbackId != null) {
            bridge.saveCall(call);
        }

        //调用api接口，发送数据到微信
        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }


    /**
     * 分享链接
     *
     * @param call
     */
    @PluginMethod
    public void shareLink(PluginCall call) {

        // String webpageUrl = call.getString("url");
        // String title = call.getString("title");
        // String description = call.getString("description");
        // String thumb = call.getString("thumb");
        // 初始化一个WXWebpageObject，填写url
        WXWebpageObject webpage = new WXWebpageObject();
        webpage.webpageUrl = call.getString("url");

        // 用 WXWebpageObject 对象初始化一个 WXMediaMessage 对象
        WXMediaMessage msg = new WXMediaMessage(webpage);
        msg.title = call.getString("title");
        msg.description = call.getString("description");

        String thumb = call.getString("thumb");
        if (thumb == null || thumb.length() <= 0) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        msg.thumbData = Util.getByteArrayThumbFromBitmap(Util.covertBase64ToBitmap(thumb));

        // 构造一个Req
        SendMessageToWX.Req req = new SendMessageToWX.Req();
        req.transaction = buildTransaction("webpage");
        req.message = msg;
        req.scene = call.getInt("scene");

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        callbackId = call.getCallbackId();
        if (callbackId != null) {
            bridge.saveCall(call);
        }

        // 调用api接口，发送数据到微信
        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }

    /**
     * 分享图片
     *
     * @param call
     */
    @PluginMethod
    public void shareImage(PluginCall call) {
        String imgUrl = call.getString("imageUrl");
        //Integer scene = call.getInt("scene");
        if (imgUrl == null || imgUrl.length() <= 0) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        String imagePath = getContext().getCacheDir().getAbsolutePath() + "/" + imgUrl;
        Bitmap bmp = BitmapFactory.decodeFile(imagePath);

        //初始化 WXImageObject 和 WXMediaMessage 对象
        WXImageObject imgObj = new WXImageObject(bmp);
        WXMediaMessage msg = new WXMediaMessage();
        msg.mediaObject = imgObj;

        //设置缩略图
        msg.thumbData = Util.getByteArrayThumbFromBitmap(bmp);

        //构造一个Req
        SendMessageToWX.Req req = new SendMessageToWX.Req();
        req.transaction = buildTransaction("img");
        req.message = msg;
        req.scene = call.getInt("scene");

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        callbackId = call.getCallbackId();
        if (callbackId != null) {
            bridge.saveCall(call);
        }

        //调用api接口，发送数据到微信
        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }

    /**
     * 分享微信小程序
     *
     * @param call
     */
    @PluginMethod
    public void shareMiniProgram(PluginCall call) {

        // String webpageUrl = call.getString("webpageUrl");
        // String userName = call.getString("userName");
        // String path = call.getString("path");
        // String hdImageData = call.getString("hdImageData");
        // Boolean withShareTicket = call.getBoolean("withShareTicket");
        // String title = call.getString("title");
        // String description = call.getString("description");
        // Integer scene = call.getInt("scene");
        // Integer miniProgramType = call.getInt("miniProgramType");
        WXMiniProgramObject miniProgramObj = new WXMiniProgramObject();
        miniProgramObj.webpageUrl = call.getString("webpageUrl");   // 兼容低版本的网页链接
        miniProgramObj.miniprogramType = call.getInt("miniProgramType");    // 正式版:0，测试版:1，体验版:2
        miniProgramObj.userName = call.getString("userName");   // 小程序原始id
        miniProgramObj.path = call.getString("path");       // 小程序页面路径；对于小游戏，可以只传入 query 部分，来实现传参效果，如：传入 "?foo=bar"
        miniProgramObj.withShareTicket = call.getBoolean("withShareTicket");
        WXMediaMessage msg = new WXMediaMessage(miniProgramObj);
        msg.title = call.getString("title");    // 小程序消息title
        msg.description = call.getString("description");    // 小程序消息desc
        String hdImageData = call.getString("hdImageData");
        if (hdImageData == null || hdImageData.length() <= 0) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        new DownloadImageTask(msg).execute(hdImageData); // 小程序消息封面图片，小于128k

        SendMessageToWX.Req req = new SendMessageToWX.Req();
        req.transaction = buildTransaction("miniProgram");
        req.message = msg;
        req.scene = SendMessageToWX.Req.WXSceneSession;  // 目前只支持会话

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        callbackId = call.getCallbackId();
        if (callbackId != null) {
            bridge.saveCall(call);
        }

        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }

    /**
     * 调起微信小程序
     *
     * @param call
     */
    @PluginMethod
    public void launchMiniProgram(PluginCall call) {

        WXLaunchMiniProgram.Req req = new WXLaunchMiniProgram.Req();
        req.userName = call.getString("userName"); // 填小程序原始id
        req.path = call.getString("path");   //拉起小程序页面的可带参路径，不填默认拉起小程序首页，对于小游戏，可以只传入 query 部分，来实现传参效果，如：传入 "?foo=bar"。
        req.miniprogramType = call.getInt("miniProgramType");// 可选打开 开发版，体验版和正式版

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        callbackId = call.getCallbackId();
        if (callbackId != null) {
            bridge.saveCall(call);
        }

        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }

    /**
     * 微信登录
     *
     * @param call
     */
    @PluginMethod
    public void sendAuthRequest(PluginCall call) {

        final SendAuth.Req req = new SendAuth.Req();
        req.scope = call.getString("scope");
        req.state = call.getString("state");

        if (!req.checkArgs()) {
            call.reject(ERROR_INVALID_PARAMETERS);
            return;
        }
        callbackId = call.getCallbackId();
        if (callbackId != null) {
            bridge.saveCall(call);
        }


        if (!wxApi.sendReq(req)) {
            call.reject(ERROR_SEND_REQUEST_FAILED);
        }
    }

    public static IWXAPI getWxAPI(Context ctx) {
        if (wxApi == null) {
            wxApi = WXAPIFactory.createWXAPI(ctx, getWxAppId(), true);
        }
        return wxApi;
    }

    /**
     * 获取微信appId
     *
     * @return
     */
    public static String getWxAppId() {
        return bridge.getConfig().getString("wechatAppId");
    }


    /**
     * 获取商户id
     *
     * @return
     */
    private String getMchId() {
        return bridge.getConfig().getString("mchid");
    }

    /**
     * 向微信注册app
     */
    private void registerWeChat() {
        getWxAPI(this.getContext()).registerApp(getWxAppId());
    }

    private String buildTransaction(final String type) {
        return (type == null) ? String.valueOf(System.currentTimeMillis()) : type + System.currentTimeMillis();
    }

    private class DownloadImageTask extends AsyncTask<String, Void, Bitmap> {
        private WXMediaMessage mediaMessage;

        public DownloadImageTask(WXMediaMessage mediaMessage) {
            this.mediaMessage = mediaMessage;
        }

        protected Bitmap doInBackground(String... urls) {
            return Util.getBitmap(urls[0]);
        }


        /*  @Override
        protected void onPreExecute() {
            // TODO Auto-generated method stub
            super.onPreExecute();
            showProgressDialog();
        }*/

        protected void onPostExecute(Bitmap bitmap) {
            if (Util.isOverSize(bitmap, 128)) {
                bitmap = Util.imageZoom(bitmap);
            }
            this.mediaMessage.thumbData = Util.bmpToByteArray(bitmap, true);
        }
    }

}

File: ios/Plugin/WechatSDKPlugin.swift
import Foundation
import Capacitor

/**
 * Please read the Capacitor iOS Plugin Development Guide
 * here: https://capacitorjs.com/docs/plugins/ios
 */
@objc(WechatSDKPlugin)
public class WechatSDKPlugin: CAPPlugin {
    private let implementation = WechatSDK()
    var callbackId: String? = nil
    public static var sharedInstance:WechatSDKPlugin? = nil
    override public func load() {
        WechatSDKPlugin.sharedInstance=self
        //在register之前打开log, 后续可以根据log排查问题 0:打印正常日志；1:打印详细日志
        //        WXApi.startLog(by: WXLogLevel(rawValue: 0)!) { log in
        //            print("WeChatSDK:",log);
        //        }
        
        //注册微信支付
        let appId = self.bridge?.config.getString("wechatAppId") ?? ""
        let universalLink = self.bridge?.config.getString("wechatUniversalLink") ?? ""
        //        print("wechat appId:",appId)
        //        print("wechat universalLink:",universalLink)
        WXApi.registerApp(appId , universalLink: universalLink );
        //调用自检函数
        //        WXApi.checkUniversalLinkReady { step, result in
        //            print("自检测步骤=====>",step, "结果=====>",result.success, "错误原因=====>",result.errorInfo,"建议=====>", result.suggestion);
        //        }
        
        //        WXApiManager.sharedInstance.delegate=self
        
    }
    
    @objc func echo(_ call: CAPPluginCall) {
        let value = call.getString("value") ?? ""
        call.resolve([
            "value": implementation.echo(value)
        ])
    }
    
    // 调起支付
    @objc func pay(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            
            guard let partnerId = self.bridge?.config.getString("mchid") else {
                call.reject("商户号没有配置.relation param: partnerId")
                return
            }
            guard let prepayId = call.getString("prepayId") else {
                call.reject("参数错误：预支付交易会话ID不能为空.relation param: prepayId")
                return
            }
            guard let nonceStr = call.getString("nonceStr") else {
                call.reject("参数错误：随机字符串不能为空.relation param: nonceStr")
                return
            }
            guard let timeStamp = call.getString("timeStamp") else {
                call.reject("参数错误：时间戳不能为空.relation param: timeStamp")
                return
            }
            guard let package = call.getString("packageValue") else {
                call.reject("参数错误：订单详情扩展字符串不能为空.relation param: packageValue")
                return
            }
            guard let sign = call.getString("sign") else {
                call.reject("参数错误：预支付交易会话ID不能为空.relation param: sign")
                return
            }
            
            let  req = PayReq()
            req.partnerId           = partnerId ;
            req.prepayId            = prepayId;
            req.nonceStr            = nonceStr;
            req.timeStamp           = UInt32(timeStamp)!
            req.package             = package;
            req.sign                = sign;
            
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            //日志输出
            debugPrint("微信支付调起url:partnerId-->",partnerId,"prepayId-->",prepayId,"nonceStr-->",nonceStr,"timeStamp-->",timeStamp,"package-->",package,"sign-->",sign,separator: " :")
            
            WXApi.send(req) {success in
                if !success{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
        }
        
    }
    
    
    // 微信是否安装
    @objc func isInstalled(_ call: CAPPluginCall) {
        //use DispatchQueue.main.async isn't necessary
        DispatchQueue.main.async {
            if !WXApi.isWXAppInstalled(){
                call.reject(WXApiManager.ERROR_WECHAT_NOT_INSTALLED,"-8");
            }else{
                call.resolve();
            }
            
        }
        
    }
    
    // 分享文本
    @objc func shareText(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            
            let req = SendMessageToWXReq()
            
            req.text = call.getString("text")!
            req.bText = true
            req.scene = Int32(call.getInt("scene")!)
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            WXApi.send(req) { (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
                
            }
            
        }
    }
    
    // 分享链接
    @objc func shareLink(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            
            let ext = WXWebpageObject()
            ext.webpageUrl = call.getString("url")!
            
            
            let message = WXMediaMessage()
            message.title = call.getString("title")!
            message.description = call.getString("description")!
            
            let thumb = call.getString("thumb") ?? ""
            
            let url = URL(string: thumb)
            if let data = try? Data(contentsOf: url!) {
                let thumbimage = self.compressImage(UIImage(data: data)!, toByte: 65536)
                message.setThumbImage(thumbimage)
            }
            
            message.mediaObject = ext
            
            let req = SendMessageToWXReq()
            req.bText = false
            req.scene = Int32(call.getInt("scene")!)
            req.message = message
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            WXApi.send(req) { (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
            
        }
    }
    
    // 分享图片
    @objc func shareImage(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            
            let ext = WXImageObject()
            
            let documentsUrl = FileManager.default.urls(for: .documentDirectory, in: .userDomainMask).first!;
            
            let image = documentsUrl.appendingPathComponent(call.getString("image")!)
            
            var thumbImage :UIImage? = nil;
            if let imageData = try? Data(contentsOf: image) {
                ext.imageData = UIImage(data: imageData)!.pngData()!
                thumbImage = self.compressImage(UIImage(data: imageData)!, toByte: 65536)
            }
            
            let message = WXMediaMessage()
            message.title = call.getString("title")!
            message.description = call.getString("description")!
            if thumbImage != nil {
                message.setThumbImage(thumbImage!)
            }
            
            message.mediaObject = ext
            
            let req = SendMessageToWXReq()
            req.bText = false
            req.scene = Int32(call.getInt("scene")!)
            req.message = message
            
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            WXApi.send(req) { (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
            
        }
    }
    
    // 分享小程序
    @objc func shareMiniProgram(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            let obj = WXMiniProgramObject()
            obj.webpageUrl = call.getString("webpageUrl")!
            obj.userName = call.getString("userName")!
            obj.path = call.getString("path")!
            
            let hdImageData = call.getString("hdImageData")!
            let url = URL(string: hdImageData)
            
            if let data = try? Data(contentsOf: url!) {
                obj.hdImageData = data;
            }
            
            obj.withShareTicket = call.getBool("withShareTicket")!
            obj.miniProgramType = WXMiniProgramType(rawValue: UInt(call.getInt("miniProgramType")!)) ?? WXMiniProgramType.release
            
            
            let message = WXMediaMessage()
            message.title = call.getString("title")!
            message.description = call.getString("description")!
            message.thumbData = nil
            message.mediaObject = obj
            
            let req = SendMessageToWXReq()
            req.bText = false
            req.scene = Int32(call.getInt("scene")!)
            req.message = message
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            WXApi.send(req) { (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
        }
    }
    
    // 调起小程序
    @objc func launchMiniProgram(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            
            let req = WXLaunchMiniProgramReq()
            req.userName = call.getString("userName")!
            req.path = call.getString("path")!
            req.miniProgramType = WXMiniProgramType(rawValue: UInt(call.getInt("miniProgramType")!)) ?? WXMiniProgramType.release
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            WXApi.send(req) { (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
            
        }
    }
    
    // 授权登录
    @objc func sendAuthRequest(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            //构造SendAuthReq结构体
            guard let scope = call.getString("scope")  else {
                call.reject("scope 参数错误")
                return
            }
            guard let state = call.getString("state")  else {
                call.reject("state 参数错误")
                return
            }
            
            let req = SendAuthReq()
            req.scope = scope;
            req.state = state;
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            
            //第三方向微信终端发送一个SendAuthReq消息结构
            WXApi.send(req){ (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
            
        }
        
    }
    
    // 拉起微信客服
    @objc func wxOpenCustomerServiceChat(_ call: CAPPluginCall) {
        DispatchQueue.main.async {
            guard let corpId = call.getString("corpId")  else {
                call.reject("corpId 参数异常")
                return
            }
            guard let url = call.getString("url")  else {
                call.reject("url 参数异常")
                return
            }
            
            let req = WXOpenCustomerServiceReq();
            req.corpid = corpId;    //企业ID
            req.url = url;          //客服URL
            
            
            guard let callbackId = call.callbackId else {
                call.reject("The call has no callbackId")
                return
            }
            self.callbackId = callbackId
            self.bridge?.saveCall(call);
            
            WXApi.send(req){ (res) in
                if !res{
                    call.reject(WXApiManager.ERROR_SEND_REQUEST_FAILED,"-9")
                }
            }
            
        }
        
    }
    
    
    @objc private func compressImage(_ image: UIImage, toByte maxLength: Int) -> UIImage {
        var compression: CGFloat = 1
        guard var data = image.jpegData(compressionQuality: compression),
              data.count > maxLength else { return image }
        
        // Compress by size
        var max: CGFloat = 1
        var min: CGFloat = 0
        for _ in 0..<6 {
            compression = (max + min) / 2
            data = image.jpegData(compressionQuality: compression)!
            if CGFloat(data.count) < CGFloat(maxLength) * 0.9 {
                min = compression
            } else if data.count > maxLength {
                max = compression
            } else {
                break
            }
        }
        var resultImage: UIImage = UIImage(data: data)!
        if data.count < maxLength { return resultImage }
        
        // Compress by size
        var lastDataLength: Int = 0
        while data.count > maxLength, data.count != lastDataLength {
            lastDataLength = data.count
            let ratio: CGFloat = CGFloat(maxLength) / CGFloat(data.count)
            let size: CGSize = CGSize(width: Int(resultImage.size.width * sqrt(ratio)),
                                      height: Int(resultImage.size.height * sqrt(ratio)))
            UIGraphicsBeginImageContext(size)
            resultImage.draw(in: CGRect(x: 0, y: 0, width: size.width, height: size.height))
            resultImage = UIGraphicsGetImageFromCurrentImageContext()!
            UIGraphicsEndImageContext()
            data = resultImage.jpegData(compressionQuality: compression)!
        }
        return resultImage
    }
    
}
